Fuzzer is broken

Running
```
make fuzz-run-parse
```
I see
```=> ERROR [ 2/17] RUN groupadd --gid 1000 prism     && useradd --uid 1000 --gid 1000 -m pris  0.2s
------
 > [ 2/17] RUN groupadd --gid 1000 prism     && useradd --uid 1000 --gid 1000 -m prism     && apt update && apt install -y sudo     && echo prism ALL=(root) NOPASSWD:ALL > /etc/sudoers.d/prism     && chmod 0440 /etc/sudoers.d/prism:
0.094 groupadd: GID '1000' already exists
------
Dockerfile:8
--------------------
   7 |
   8 | >>> RUN groupadd --gid $USER_GID $USERNAME \
   9 | >>>     && useradd --uid $USER_UID --gid $USER_GID -m $USERNAME \
  10 | >>>     && apt update && apt install -y sudo \
  11 | >>>     && echo $USERNAME ALL=\(root\) NOPASSWD:ALL > /etc/sudoers.d/$USERNAME \
  12 | >>>     && chmod 0440 /etc/sudoers.d/$USERNAME
  13 |
--------------------
ERROR: failed to build: failed to solve: process "/bin/sh -c groupadd --gid $USER_GID $USERNAME     && useradd --uid $USER_UID --gid $USER_GID -m $USERNAME     && apt update && apt install -y sudo     && echo $USERNAME ALL=\\(root\\) NOPASSWD:ALL > /etc/sudoers.d/$USERNAME     && chmod 0440 /etc/sudoers.d/$USERNAME" did not complete successfully: exit code: 4
make: *** [fuzz-docker-build] Error 1
```
I believe that further along in the process there will be other issues e.g. AFL_HARDEN is now mutually exclusive with ASAN.

I have a [fork](https://github.com/stevenjohnstone/prism/blob/sjj/fix-fuzzing/docs/fuzzing.md) where I've fixed the build issues and made a series of improvements to make it easier to run the fuzzer, monitor it, and triage the results. Note that there are some tentative fixes applied to that branch for some issues detected during fuzzing (out-of-bounds reads and some undefined behaviour).

--- a/Makefile
+++ b/Makefile
@@ -63,12 +63,12 @@ build/fuzz.%: $(SOURCES) fuzz/%.c fuzz/fuzz.c
 	$(ECHO) "building $* fuzzer"
 	$(Q) $(MAKEDIRS) $(@D)
 	$(ECHO) "building main fuzz binary"
-	$(Q) AFL_HARDEN=1 afl-clang-lto $(DEBUG_FLAGS) $(CPPFLAGS) $(CFLAGS) $(FUZZ_FLAGS) -O0 -fsanitize-ignorelist=fuzz/asan.ignore -fsanitize=fuzzer,address -ggdb3 -std=c99 -Iinclude -o $@ $^
+	$(Q) afl-clang-lto $(DEBUG_FLAGS) $(CPPFLAGS) $(CFLAGS) $(FUZZ_FLAGS) -O0 -fsanitize-ignorelist=fuzz/asan.ignore -fsanitize=fuzzer,address -ggdb3 -std=c99 -Iinclude -o $@ $^
 	$(ECHO) "building cmplog binary"
-	$(Q) AFL_HARDEN=1 AFL_LLVM_CMPLOG=1 afl-clang-lto $(DEBUG_FLAGS) $(CPPFLAGS) $(CFLAGS) $(FUZZ_FLAGS) -O0 -fsanitize-ignorelist=fuzz/asan.ignore -fsanitize=fuzzer,address -ggdb3 -std=c99 -Iinclude -o $@.cmplog $^
+	$(Q) AFL_LLVM_CMPLOG=1 afl-clang-lto $(DEBUG_FLAGS) $(CPPFLAGS) $(CFLAGS) $(FUZZ_FLAGS) -O0 -fsanitize-ignorelist=fuzz/asan.ignore -fsanitize=fuzzer,address -ggdb3 -std=c99 -Iinclude -o $@.cmplog $^
 
 build/fuzz.heisenbug.%: $(SOURCES) fuzz/%.c fuzz/heisenbug.c
-	$(Q) AFL_HARDEN=1 afl-clang-lto $(DEBUG_FLAGS) $(CPPFLAGS) $(CFLAGS) $(FUZZ_FLAGS) -O0 -fsanitize-ignorelist=fuzz/asan.ignore -fsanitize=fuzzer,address -ggdb3 -std=c99 -Iinclude -o $@ $^
+	$(Q) afl-clang-lto $(DEBUG_FLAGS) $(CPPFLAGS) $(CFLAGS) $(FUZZ_FLAGS) -O0 -fsanitize-ignorelist=fuzz/asan.ignore -fsanitize=fuzzer,address -ggdb3 -std=c99 -Iinclude -o $@ $^
 
 fuzz-debug:
 	$(ECHO) "entering debug shell"

--- a/fuzz/docker/Dockerfile
+++ b/fuzz/docker/Dockerfile
@@ -1,8 +1,9 @@
-FROM aflplusplus/aflplusplus
+FROM aflplusplus/aflplusplus:v4.32c
 
 ARG USERNAME=prism
 ARG USER_UID=1000
 ARG USER_GID=$USER_UID
+ARG RUBY_VERSION=3.3.10
 ENV MAKEFLAGS=-j8
 
 RUN groupadd --gid $USER_GID $USERNAME \
@@ -12,10 +13,10 @@ RUN groupadd --gid $USER_GID $USERNAME \
     && chmod 0440 /etc/sudoers.d/$USERNAME
 
 
-RUN wget https://cache.ruby-lang.org/pub/ruby/3.2/ruby-3.2.2.tar.gz
-RUN tar -xvf ruby-3.2.2.tar.gz
+RUN wget https://cache.ruby-lang.org/pub/ruby/3.3/ruby-${RUBY_VERSION}.tar.gz -O ruby.tar.gz
+RUN mkdir ruby-source && tar -xvf ruby.tar.gz -C ruby-source --strip-components=1
 RUN apt update && apt -y install libyaml-dev libz-dev libssl-dev
-RUN cd ruby-3.2.2 && ./configure --disable-install-doc && make && make install
+RUN cd ruby-source && ./configure --disable-install-doc && make && make install
 RUN gem install rake-compiler ruby_memcheck
 RUN git clone https://github.com/pwndbg/pwndbg && cd pwndbg && ./setup.sh
 ENV LC_CTYPE=C.UTF-8

